kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/ScanSuite.scala

```diff
@@ -15,29 +15,955 @@
+        greaterThan(col("c4"), ofFloat(1.0f)),
+        equals(col("c6"), ofDate(InternalUtils.daysSinceEpoch(Date.valueOf("2002-02-02")))),
+        greaterThan(col("c9"), ofDecimal(JBigDecimal.valueOf(1.5), 3, 2)),
+        // We don't currently skip for timestamps but this will still be a hit once we do
+        getTimestampPredicate(">=", col("c5"), "2001-01-01T01:00:00-07:00"),
+        // Once we skip for timestamps this should be a miss
+        getTimestampPredicate(">=", col("c5"), "2003-01-01T01:00:00-07:00")
+      ),
+      misses = Seq(
+        equals(col("c1"), ofInt(10)),
+        lessThan(col("c3"), ofFloat(0.5f)),
+        greaterThan(col("c4"), ofFloat(5.0f)),
+        equals(col("c6"), ofDate(InternalUtils.daysSinceEpoch(Date.valueOf("2003-02-02")))),
+        greaterThan(col("c9"), ofDecimal(JBigDecimal.valueOf(2.5), 3, 2))
+      )
+    )
+  }
+
+  test("data skipping by partition and data values - nulls") {
+    withTempDir { tableDir =>
+      val dataSeqs = Seq( // each sequence produce a single file
+        Seq((null, null)),
+        Seq((null, "a")),
+        Seq((null, "b")),
+        Seq(("a", "a"), ("a", null)),
+        Seq(("b", null))
+      )
+      dataSeqs.foreach { seq =>
+        seq.toDF("key", "value").coalesce(1)
+          .write.format("delta").partitionBy("key").mode("append").save(tableDir.getCanonicalPath)
+      }
+      def checkResults(
+          predicate: Predicate, expNumPartitions: Int, expNumFiles: Long): Unit = {
+        val snapshot = latestSnapshot(tableDir.getCanonicalPath)
+        val scanFiles = collectScanFileRows(
+          snapshot.getScanBuilder(defaultTableClient)
+            .withFilter(defaultTableClient, predicate)
+            .build())
+        assert(scanFiles.length == expNumFiles,
+          s"Expected $expNumFiles but found ${scanFiles.length} for $predicate")
+
+        val partitionValues = scanFiles.map { row =>
+          InternalScanFileUtils.getPartitionValues(row)
+        }.distinct
+        assert(partitionValues.length == expNumPartitions,
+          s"Expected $expNumPartitions partitions but found ${partitionValues.length}")
+      }
+
+      // Trivial base case
+      checkResults(
+        predicate = AlwaysTrue.ALWAYS_TRUE,
+        expNumPartitions = 3,
+        expNumFiles = 5)
+
+      // Conditions on partition key
+      checkResults(
+        predicate = isNotNull(col("key")),
+        expNumPartitions = 2,
+        expNumFiles = 2) // 2 files with key = 'a', and 1 file with key = 'b'
+
+      checkResults(
+        predicate = equals(col("key"), ofString("a")),
+        expNumPartitions = 1,
+        expNumFiles = 1) // 1 files with key = 'a'
+
+
+      checkResults(
+        predicate = equals(col("key"), ofString("b")),
+        expNumPartitions = 1,
+        expNumFiles = 1) // 1 files with key = 'b'
+
+      // TODO shouldn't partition filters on unsupported expressions just not prune instead of fail?
```

Comment:
@vkorukanti what do you think? It seems like it should just be pushed back to the connector as remainingPredicate

Reply:
Returning as the remaining predicate makes sense. Let me know, I can make a PR for this. We may need to split the partition predicate (if combined with AND) to figure out which parts we can use using the expression handler.